### DISCUSSION TITLE
Why use memory vs call data

### DISCUSSION THREAD

**Discussion by cursedId0l** [Cyfrin/foundry-full-course-cu#36] [Q&A] [answered]
I'm currently studying the use of memory and calldata in Solidity, as taught by Patrick. I understand that a variable declared with memory can be changed, while one declared with calldata cannot.

In the example below from the course, a memory keyword is used for a variable that isn't being redefined:

```javascript
function addPerson(string memory _name, uint256 _favNum) public {
   myFriends.push(Person(_favNum,_name)); 
}
```

Given that the variable _name isn't being changed in the function, why was memory chosen instead of calldata? Maybe I'm just using my javascript brain, but it seems akin to declaring a variable with let when const would be more appropriate. Could you explain why memory seems to be the default choice here?

**Comment by cromewar** [comment]
Hello @jtardioli  you are right on your understanding between the difference between `memory` and `calldata`, to answer your questions:

memory is chosen for _name for the reason that the _name argument comes from an external caller, and we want to create a temporary copy of it in the EVM memory. By declaring _name as a memory type, you're saying "_name is a new temporary variable, and its initial value is to be copied from the function call data."

On the other hand, calldata is a special data location and it's cheaper in terms of gas cost since calldata is read-only. However, calldata is only available for external function's parameters. An external function is part of the contract's interface, which means it can be called from other contracts and transactions. If you have a public function like addPerson, then you can't use calldata.

Also, it's important to note that while the memory keyword doesn't make a variable immutable like const would in JavaScript, Solidity does have its own keyword for making variables unchangeable: constant or immutable. However, these keywords are typically used for state variables, not function parameters.

@PatrickAlphaC did I miss anything?

**Reply by cursedId0l** [reply]
Thanks so much, very helpful reply!

**Comment by PatrickAlphaC** [comment, marked as answer]
I used `memory` there because I wanted you to get familiar with that syntax, as in the real world that's what you'll see most often.

However, you're 100% correct! We could use `calldata` here, and that would be more gas efficient! Consider the following contract:

```javascript
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

contract StringPlay {
    string public hello;
    function saveHelloCalldata(string calldata _greeting) public {
        hello = _greeting;
    }

    function saveHelloMemory(string memory _greeting) public {
        hello = _greeting;
    }
}
```

You'll see that the `calldata` one is cheaper gas-wise. Excellent question and observation!

**Reply by cursedId0l** [reply]
Thanks Patrick!

**Comment by Anurag0103m** [comment]
Hello everyone  I am new in coding and  in blockchain too . I have here a that why we typing code twice in the above syntax ?

**Reply by cromewar** [reply]
Can you be more specific on your question @Anurag0103m?

**Comment by MolotovMM** [comment]
ty!
